packages/flutter/lib/src/widgets/scroll_notification_observer.dart

```diff
@@ -0,0 +1,174 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:collection';
+
+import 'package:flutter/foundation.dart';
+
+import 'framework.dart';
+import 'notification_listener.dart';
+import 'scroll_notification.dart';
+
+/// A [ScrollNotification] listener for [ScrollNotificationObserver].
+///
+/// [ScrollNotificationObserver] is similar to
+/// [NotificationListener]. It supports a listener list instead of
+/// just a single listener and its listeners run unconditionally, they
+/// do not require a gating boolean return value.
+typedef ScrollNotificationCallback = void Function(ScrollNotification notification);
+
+class _ScrollNotificationObserverScope extends InheritedWidget {
+  const _ScrollNotificationObserverScope({
+    Key? key,
+    required Widget child,
+    required ScrollNotificationObserverState scrollNotificationObserverState,
+  }) : _scrollNotificationObserverState = scrollNotificationObserverState,
+      super(key: key, child: child);
+
+  final ScrollNotificationObserverState  _scrollNotificationObserverState;
+
+  @override
+  bool updateShouldNotify(_ScrollNotificationObserverScope old) => _scrollNotificationObserverState != old._scrollNotificationObserverState;
+}
+
+class _ListenerEntry extends LinkedListEntry<_ListenerEntry> {
+  _ListenerEntry(this.listener);
+  final ScrollNotificationCallback listener;
+}
+
+/// The listener list state for a [ScrollNotificationObserver] returned by
+/// [ScrollNotificationObserver.of].
+///
+/// [ScrollNotificationObserver] is similar to
+/// [NotificationListener]. It supports a listener list instead of
+/// just a single listener and its listeners run unconditionally, they
+/// do not require a gating boolean return value.
+class ScrollNotificationObserverState extends State<ScrollNotificationObserver> {
```

Comment:
Nit: Should the state come after the `ScrollNotificationObserver` in the file widget to follow convention

Reply:
Yes, done.